lazyproto/decode.go

```diff
@@ -0,0 +1,186 @@
+// to use [Unmarshal] instead.
+func Decode(data []byte, def *Def) (res DecodeResult, err error) {
+	if len(data) == 0 || def == nil || len(def.m) == 0 {
+		return emptyResult, nil
+	}
+	for dec := csproto.NewDecoder(data); dec.More(); {
+		tag, wt, err := dec.DecodeTag()
+		if err != nil {
+			return emptyResult, err
+		}
+		dv, want := def.Get(tag)
+		if !want {
+			if _, err := dec.Skip(tag, wt); err != nil {
+				return emptyResult, err
+			}
+			continue
+		}
+		switch wt {
+		case csproto.WireTypeVarint, csproto.WireTypeFixed32, csproto.WireTypeFixed64:
+			// varint, fixed32, and fixed64 could be multiple Go types so
+			// grab the raw bytes and defer interpreting them to the consumer/caller
+			// . varint -> int32, int64, uint32, uint64, sint32, sint64, bool, enum
+			// . fixed32 -> int32, uint32, float32
+			// . fixed64 -> int32, uint64, float64
+			val, err := dec.Skip(tag, wt)
+			if err != nil {
+				return emptyResult, err
+			}
+			fd, err := res.getOrAddFieldData(tag, wt)
+			if err != nil {
+				return emptyResult, err
+			}
+			// Skip() returns the entire field contents, both the tag and the value, so we need to skip past the tag
+			val = val[csproto.SizeOfTagKey(tag):]
+			fd.data = append(fd.data, val)
+		case csproto.WireTypeLengthDelimited:
+			val, err := dec.DecodeBytes()
+			if err != nil {
+				return emptyResult, err
+			}
+			if subDef, ok := dv.(*Def); ok && len(subDef.m) > 0 {
+				// recurse
+				subResult, err := Decode(val, subDef)
+				if err != nil {
+					return emptyResult, err
+				}
+				fd, err := res.getOrAddFieldData(tag, wt)
+				if err != nil {
+					return emptyResult, err
+				}
+				fd.data = append(fd.data, subResult.m)
+			} else {
+				fd, err := res.getOrAddFieldData(tag, wt)
+				if err != nil {
+					return emptyResult, err
+				}
+				fd.data = append(fd.data, val)
+			}
+		default:
+			return emptyResult, fmt.Errorf("read unknown/unsupported protobuf wire type (%v)", wt)
+		}
+	}
+	return res, nil
+}
+
+// DecodeResult holds a (possibly nested) mapping of integer field tags to FieldData instances
+// which can be used to retrieve typed values for specific Protobuf message fields.
+type DecodeResult struct {
+	m map[int]*FieldData
+}
+
+// The FieldData method returns a FieldData instance for the specified tag "path", if it exists.
+//
+// The tags parameter is a list of one or more integer field tags that act as a "path" to a particular
+// field to support retreiving fields from nested messages.  Each value is used to retreieve the field
+// data at the corresponding level of nesting, i.e. a value of [1, 2] would return the field data for
+// tag 2 within the nested data for tag 1 at the root.
+//
+// For a Protobuf definition like:
+//
+//	// example.proto
+//	message Example {
+//		int64  id     = 1;
+//		Nested nested = 2;
+//	}
+//	message Nested {
+//		string name        = 1;
+//		string description = 2;
+//	}
+//
+// the fields of the Nested message could be retrieved using:
```

Comment:
this rest of this comment may be better suited in a `doc.go` file. Food for thought

Reply:
Done.  See `ExampleDecodeResult_FieldData()` in `lazyproto/decode_test.go`

Reply:
beautiful 👍